cli/tools/repl/editor.rs

```diff
@@ -391,3 +396,33 @@ impl ReplEditor {
     Ok(())
   }
 }
+
+/// A custom tab key event handler
+/// It uses a heuristic to determine if the user is requesting completion or if they want to insert an actual tab
+/// The heuristic goes like this:
+///   - If the last character before the cursor is whitespace, the the user wants to insert a tab
+///   - Else the user is requesting completion
+struct TabEventHandler;
+impl ConditionalEventHandler for TabEventHandler {
+  fn handle(
+    &self,
+    evt: &Event,
+    n: RepeatCount,
+    _: bool,
+    ctx: &EventContext,
+  ) -> Option<Cmd> {
+    debug_assert_eq!(*evt, Event::from(KeyEvent::from('\t')));
+    if ctx.line().is_empty()
+      || ctx.line()[..ctx.pos()]
```

Comment:
To save other reviewers time, `ctx.pos()` is indeed byte position https://docs.rs/rustyline/9.1.2/rustyline/struct.EventContext.html